pjsip/src/pjsip/sip_transport_tcp.c

```diff
@@ -1507,6 +1550,20 @@ static pj_bool_t on_connect_complete(pj_activesock_t *asock,
     return PJ_TRUE;
 }
 
+/* Transport connection timeout timer callback */
+static void tcp_connect_timer(pj_timer_heap_t *th, pj_timer_entry *e)
+{
+    struct tcp_transport *tcp = (struct tcp_transport*) e->user_data;
+
+    PJ_UNUSED_ARG(th);
+
+    /* Fake up a failed connection complete event from the lower layers.
+       on_connect_complete will close the socket, which will tear
+       everything down. */
+    PJ_LOG(4,(tcp->base.obj_name, "TCP transport connect timeout timed out"));
```

Comment:
Would it be safer to have a defensive "if (tcp->has_pending_connect) {" condition here to save us from unexpected pops from code paths we haven't spotted accidentally killing perfectly good connections?